src/git.rs

```diff
@@ -93,32 +95,96 @@ pub enum InitError {
+    // However, NixOS tend to have 'GlobalKnownHostsFile' in '/etc/ssh/config' that may point
+    // to files in '/nix/store', and additionally, home-manager may set 'UserKnownHostsFile' in
+    // ~/.ssh/config.
+    // We need to make sure that 'certificate_check' callback checks these files.
+    callbacks.certificate_check(move |cert, host| {
+        //return Ok(git2::CertificateCheckStatus::CertificateOk);
+        let Some(hostkey) = cert.as_hostkey().and_then(|h| h.hostkey()) else {
+            return Ok(git2::CertificateCheckStatus::CertificatePassthrough);
+        };
+        // Check local ssh config;
+        let mut known_hosts_files: Vec<String> = Vec::new();
+        if let Some(conf) = &state.local_ssh_config {
+            if let Some(files) = conf
+                .query(host)
+                .ignored_fields
+                .get(&Field::UserKnownHostsFile)
+            {
+                known_hosts_files.append(&mut files.to_owned());
+            }
+        };
+        if let Some(conf) = &state.global_ssh_config {
+            if let Some(files) = conf
+                .query(host)
+                .ignored_fields
+                .get(&Field::GlobalKnownHostsFile)
+            {
+                known_hosts_files.append(&mut files.to_owned());
+            }
+        };
+        let err_msg = match Session::new() {
+            Ok(sess) => match sess.known_hosts() {
+                Ok(mut known_hosts) => {
+                    for known_host_file in known_hosts_files {
+                        let _ = known_hosts.read_file(
+                            &PathBuf::from(known_host_file),
+                            ssh2::KnownHostFileKind::OpenSSH,
+                        );
+                    }
+                    match known_hosts.check(host, hostkey) {
+                        CheckResult::Match => {
+                            return Ok(git2::CertificateCheckStatus::CertificateOk)
+                        }
+                        CheckResult::Failure => {
+                            format!("Something prevented the check to be made")
+                        }
+                        CheckResult::Mismatch => {
+                            format!("Host was found, but the keys didn't match")
+                        }
+                        CheckResult::NotFound => {
+                            return Ok(git2::CertificateCheckStatus::CertificatePassthrough)
+                        }
+                    }
+                }
+                Err(err) => format!("Failed to load known hosts {0}", err),
+            },
+            Err(err) => format!("Failed to initialize SSH session {0}", err),
+        };
+        return Err(git2::Error::new(
+            git2::ErrorCode::GenericError,
+            git2::ErrorClass::Callback,
+            err_msg,
+        ));
+    });
+    callbacks
+}
```

Comment:
I'm terribly sorry, but frankly it's a little :ssc: :nabros: I've tried to clean it up a bit...

```suggestion
fn callbacks(state: &UpdateState) -> git2::RemoteCallbacks {
    let mut callbacks = RemoteCallbacks::new();
    callbacks
        .credentials(|_url, username, _| git2::Cred::ssh_key_from_agent(username.unwrap_or("git")))
        .certificate_check(move |cert, host| {
            // libgit2 only considers "~/.ssh/known_hosts" when checking the git host certificate,
            // see https://github.com/libgit2/libgit2/blob/115db540cfb633c2a618aa60757454839047eadf/src/libgit2/transports/ssh_libssh2.c#L435
            // However, NixOS tend to have 'GlobalKnownHostsFile' in '/etc/ssh/config' that may point
            // to files in '/nix/store', and additionally, home-manager may set 'UserKnownHostsFile' in
            // ~/.ssh/config.
            // We need to make sure that 'certificate_check' callback checks these files.
            let Some(hostkey) = cert.as_hostkey().and_then(CertHostkey::hostkey) else {
                return Ok(git2::CertificateCheckStatus::CertificatePassthrough);
            };
            // Check local ssh config;
            let get_host_files_from_field = |f: Field| -> Vec<String> {
                let Some(ref conf) = state.local_ssh_config else {
                    return Vec::new();
                };
                let mut host_params = conf.query(host);
                // NB: we own host_params, hence we can safely take out the field
                // value instead of cloning it
                host_params.ignored_fields.remove(&f).unwrap_or_default()
            };
            let known_hosts_files: Vec<String> =
                get_host_files_from_field(Field::UserKnownHostsFile)
                    .into_iter()
                    .chain(get_host_files_from_field(Field::GlobalKnownHostsFile))
                    .collect();
            fn mk_err(err_msg: impl AsRef<str>) -> git2::Error {
                git2::Error::new(
                    git2::ErrorCode::GenericError,
                    git2::ErrorClass::Callback,
                    err_msg,
                )
            }
            let sess = Session::new()
                .map_err(|err| mk_err(format!("Failed to initialize SSH session {0}", err)))?;
            let mut known_hosts = sess
                .known_hosts()
                .map_err(|err| mk_err(format!("Failed to load known hosts {0}", err)))?;
            for known_host_file in known_hosts_files {
                let _ = known_hosts.read_file(
                    &PathBuf::from(known_host_file),
                    ssh2::KnownHostFileKind::OpenSSH,
                );
            }
            match known_hosts.check(host, hostkey) {
                CheckResult::Match => Ok(git2::CertificateCheckStatus::CertificateOk),
                CheckResult::NotFound => Ok(git2::CertificateCheckStatus::CertificatePassthrough),
                CheckResult::Failure => Err(mk_err("Something prevented the host check")),
                CheckResult::Mismatch => Err(mk_err("Host was found, but the keys didn't match")),
            }
        });
    callbacks
}
```